### PR TITLE
Added addition/subtraction for Rays and Segments instead of Complement/Unions sets

### LINKED ISSUE #12988: difference/sum of Ray and Segment gives Complement/Union set
While Point + Point is defined, Ray + Ray is not (in a vector sense) and it seems like that might be useful. Instead, current behavior gives a Complement set. (And also does so for the difference of Segments, though perhaps that should just raise an error.) And sums give Unions. 

```
>>> Ray((1,1),(2,2))-Ray((2,3),(4,5))
Ray2D(Point2D(1, 1), Point2D(2, 2)) \ Ray2D(Point2D(2, 3), Point2D(4, 5))
>>> Segment((1,1),(2,2))-Segment((2,3),(4,5))
Segment2D(Point2D(1, 1), Point2D(2, 2)) \ Segment2D(Point2D(2, 3), Point2D(4, 5))
>>> Segment((1,1),(2,2))+Segment((2,3),(4,5))
Union(Segment2D(Point2D(1, 1), Point2D(2, 2)), Segment2D(Point2D(2, 3), Point2D(4, 5)))
```

### PR DESCRIPTION
**Pull request by srihegde**
Fixes #12988.
For Ray class, I have handled the exception of addition(subtraction) of opposite(same) rays.
Added some tests for 2D case. This could be extended to 3D as well.

### CODE REVIEW

**Comment by smichr** [comment]
What is the rationale for using unit distance with Ray but not with Segment? (Note also that r.direction.unit is the way to get a unit length direction.)

**Comment by smichr** [comment]
I doub't it is a good idea to define segment addition since segments do not have a direction. And I would expect r1 and r2 (rays) to be added vectorially like `Ray(r1.p1, r2.p2 + r2.direction)`

**Comment by srihegde** [comment]
**Reply to comment 1**
Thanks for point out about 'unit'. I used unit distance with Rays only because they do not have a sense of magnitude(as their magnitude is infinity) but only direction. I have pointed it out below with an example.

**Reply to comment 2**
I assumed the segments to have direction. But if they don't have direction then it's a meaningless operation on segments. 
Did you mean `Ray(r1.p1, r1.p2 + r2.direction)`? If so then this does not take care of magnitude issues (For eg- Take `r1 = [(0,1),(0,3)] and r2 = [(1,0),(3,0)])`). Thus it needs to be normalized.

Thanks for the valuable suggestions!

**Comment by smichr** [comment]
> because they do not have a sense of magnitude

I can't believe I have been ignoring this. I have been thinking about them like a vector. So segments don't have direction and rays don't have magnitude. It's probably not right to grant either one the property that they don't have. I think it's better to just raise a TypeError when these operations are attempted, something like:
```
raise TypeError("unsupported operand type(s) for +: '%s' and '%s'" % (type(arg1), type(arg2)))
```

And yes -- thinking of vectors -- I intended the addition to be as I gave it. It is easy enough for someone to do as I did, however, if they want to add rays as vectors:
```
def direction(rays):
  "return sum of directions for give rays"
  result = Point(0,0) 
  for r in rays:
    result = result + r.direction
  return result
```

**Comment by srihegde** [comment]
We can absolutely raise a TypeError for such operations in Segments. Thanks for the suggestion.

I could not understand how `Ray(r1.p1, r2.p2 + r2.direction)` correctly calculates the resultant. As per my calculation I am getting correct results with:

- ` r_dir = r1.direction.unit + r2.direction.unit` -> This computes the direction of resultant ray whose source is at origin.
- `r = Ray(r1.p1, r1.p1+r_dir)` -> This simply does a translation to the source to `r1.p1` from origin.

**Comment by smichr** [comment]
> I could not understand how Ray(r1.p1, r2.p2 + r2.direction) correctly calculates the resultant

It doesn't. It should be `r1.p2 + r2.direction`. This avoids an extra call to direction. They are the same:

```
>>> from random import randint
>>> 
>>> def pt():
...     return randint(-100,100), randint(-100,100)
>>>
>>> for i in range(10):
...     r1 = Ray(pt(), pt())
...     r2 = Ray(pt(), pt())
...     d1 = r1.direction
...     d2 = r2.direction
...     assert Ray(r1.p1, r1.p2 + d2) == Ray(r1.p1, r1.p1+d1+d2)
```

**Comment by srihegde** [comment]
Thanks for the clarification. I hope my last commit closes this issue.

**Comment by smichr** [comment]
OK, I understand what your direction.unit result gives: it is the bisector of the two rays.

**Comment by smichr** [comment]
There is no unique way to give the result. Here are 3 equivalent ways to add r1 and r2:

```
>>> r1
Ray2D(Point2D(0, 0), Point2D(1, 2))
>>> r2
Ray2D(Point2D(0, 0), Point2D(1, 0))

>>> r3
Ray2D(Point2D(0, 0), Point2D(1, -1/2 + sqrt(5)/2))
>>> r4
Ray2D(Point2D(0, 0), Point2D(sqrt(5)/5 + 1, 2*sqrt(5)/5))
>>> ra
Ray2D(Point2D(0, 0), Point2D(1, tan(acos(sqrt(5)/5)/2)))

>>> r3.slope.n()
0.618033988749895
>>> r4.slope.n()
0.618033988749895
>>> ra.slope.n()
0.618033988749895
```

Should one form be favored over another?

**Comment by srihegde** [comment]
That's absolutely correct. We can either report p1 and direction(slope) of the resultant or a tuple of p1 and p2 such that p2 is unit distance away from p1 in the direction of resultant ray.
Presently the code does not normalize the resultant ray and produces p2 directly from sum (of 2 unit rays).

**Comment by smichr** [comment]
> Which of the 2 is better in this case?

We are trying to do something that I think shouldn't be done. I think Segment and Ray operations (at least addition and subtraction) should just raise an error: " I think it's better to just raise a TypeError when these operations are attempted". The problem with Ray is that there is no canonical right answer: should the addition create a weighted average based on the p1 and p2 or an unweighted average (the bisector)?

If Segment becomes ordered (instead of unordered) it might make sense to arbitrarily assign addition like vector addition.

**Comment by srihegde** [comment]
Thanks for the clarification. Updated.

File: sympy/geometry/tests/test_line.py [hunk 1]
```diff
@@ -53,6 +53,14 @@ def test_angle_between():
                                 Line3D(Point3D(0, 0, 0), Point3D(5, 0, 0))), acos(sqrt(3) / 3)
 
 
+def test_closing_angle():
```
Comment (smichr) [inline sympy/geometry/tests/test_line.py]: I wonder why this is showing up here when it was add [already](https://github.com/sympy/sympy/pull/13002/files).

**Comment by smichr** [comment]
Wondering why operations are defined at all, e.g. `Line((0,0),slope=1)**2` works. I tracked it down to the fact that LinearEntity derives from GeometrySet. So perhaps someone thought that representing geometrical entities as sets of points was a good idea. I don't know why one would do this instead of using sets, especially image sets. Perhaps it would be better to give geometrical entities an `as_set` method that returns the appropriate Set.

```
>>> source(GeometrySet)
In file: /sympy/sympy/geometry/entity.py
class GeometrySet(GeometryEntity, Set):
    """Parent class of all GeometryEntity that are also Sets
    (compatible with sympy.sets)
```

This PR should probably be closed without committing in light of this.

We could also track down the person that decided on this inheritance and ask them about the rationale.

**Comment by srihegde** [comment]
Sure. Sounds good.

File: sympy/geometry/tests/test_line.py (hunk 1 above)
Comment (srihegde) [inline sympy/geometry/tests/test_line.py]: Nothing to worry about I think. I took this from the latest version of test_line.py of master as I had some version issues with my local copy of this file.

Comment (srihegde) [inline sympy/geometry/tests/test_line.py]: Nothing to worry about I think. I took this from the latest version of test_line.py of master as I had some version issues with my local copy of this file.

**Comment by smichr** [comment]
ping @siefkenj , can you help us understand what I observed [here](https://github.com/sympy/sympy/pull/13020#issuecomment-318151664) regarding GeometrySet?

**Comment by smichr** [comment]
> I don't see how inheriting from ImageSet would be any better.  What advantages would that have?  (I could see an as_imageset method though).

Not inheriting...just having the method to produce a set. The current problem is that something like a segment squared doesn't do anything and `(Segment((0,0),(1,0))**2).area` just raises an error.

**Comment by siefkenj** [comment]
It would be nice for `(Segment((0,0),(1,0))**2).area` to work.

**Comment by czgdp1807** [comment]
@siefkenj Are you still interested to work on this PR? This PR looks in a pretty good state. Thanks for the contribution.
@smichr Any updates from your side on this PR. Thanks.

**Comment by siefkenj** [comment]
@czgdp1807 I assume you meant to ask @srihegde about finishing the PR?

**Comment by czgdp1807** [comment]
Yes, I asked about the status to @srihegde. Probably I clicked on your github username when github showed suggestions while tagging @srihegde

**Comment by czgdp1807** [comment]
It looks like the code in `master` what this PR intends to do,
```python
>>> a = Ray((0, 0), angle=0)
>>> b = Ray((1, 2), angle=pi/2)
>>> a.closing_angle(b)
-π 
───
 2 
>>> b.closing_angle(a)
π
─
2
>>> a.closing_angle(a)
0
```
For now closing this but if you think that this can be re-opened, just comment below or re-open it if you have the access. Thanks for the contributions.

**Comment by czgdp1807** [comment]
It looks like the original issue is still not fixed. Re-opening it.

**Pull request closed without merging**